Fast Break

HouseOfHufflepuff opened the pull request:
Fast Break is a game built on the NBA Top Shot contract. Players of Fast Break submit moments in their Top Shot collection to play the game. Each Fast Break contains n FastBreakStatistic the wallet must match or exceed.

When a player submits a lineup to Fast Break, a game token is minted to their Fast Break collection. This game token contains their submission data and links to the game. As NBA games are played, the metadata of the token (points & win) are updated. 

A Fast Break Run is a block of Fast Breaks. Winning a Fast Break puts you on the Fast Break Run leaderboard.

The resource FastBreakDaemon is reserved for the game oracle. The oracle is responsible for bringing trusted NBA data into its platform, computing points & wins, and updating the blockchain with the derived data. The FastBreakDaemon keeps the on-chain & off-chain games in sync.

Custodial submissions to Fast Break will be made by the game oracle in batch. Non-custodial submissions can be made by any wallet with Top Shots.

<img width="1224" alt="Screen Shot 2023-12-17 at 7 05 20 PM" src="https://github.com/dapperlabs/nba-smart-contracts/assets/896651/6ce1ea68-b94e-4451-a180-f7b53124dae4">

joshuahannan reviewed (commented):
Can you add comments to all your types, fields, and functions in the contract? That would make the review process much easier for us. Thank you!

HouseOfHufflepuff commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,606 @@
+    Author: Jeremy Ahrens jer.ahrens@dapperlabs.com
+*/
+
+import NonFungibleToken from 0xNFTADDRESS
+import TopShot from 0xTOPSHOTADDRESS
```
Yes, absolutely. I am doing this PR as part of a side project in my free time, but will prio it with @annatangzhao .

joshuahannan replied:
It shouldn't take much time, it is basically just a find and replace on all the contracts, transactions and scripts, then just updating the `contracts` and `templates` packages to use the new ones instead. I created an issue: https://github.com/dapperlabs/nba-smart-contracts/issues/230

turbolent commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+        }
+
+        pub fun isWinner(): Bool {
+            let fastBreak : FastBreak.FastBreakGame = FastBreak.fastBreakGameByID[self.fastBreakGameID]!
+            let submission : FastBreak.FastBreakSubmission = fastBreak.submissions[self.mintedTo]!
```
Here and elsewhere: No need for type annotations, simply use type inference: 
```suggestion
            let fastBreak = FastBreak.fastBreakGameByID[self.fastBreakGameID]!
            let submission = fastBreak.submissions[self.mintedTo]!
```

turbolent commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+    /// Metadata Dictionaries
+    ///
+    access(self) let fastBreakRunByID:        {String: FastBreakRun}
+    access(self) let fastBreakGameByID:           {String: FastBreakGame}
```
No idea about the business logic, so may not apply/be possible: Maybe consider refactoring the architecture of how/where games are stored. Currently, the contract itself stores all games, which might become a scaling bottleneck. Is this really necessary?

Maybe try and see if account-based storage architecture could work here. For example, for a game with a "tournament"/"round", allow it to be created and stored in any account.

joshuahannan replied:
agreed with this. It might be useful for users to store the games they participate in in their own account or something like that

HouseOfHufflepuff replied:
Yes, I agree and that's what I was going to do initially. It's more ideal for sure.

The problem that I could not figure out how to solve is that the game score for each user will be updated 100s of times by the oracle as the NBA games are going on. I can write to the token for custodial accounts I have the key for but don't think I can write to a non-custodial account token. Non-custodial accounts would not be able to get their token scored, afaik.

I don't write a ton of cadence so perhaps I am missing something, but wanted to explain why I took the direction of game object storage in the contract as opposed to wallets.

tpetrychyn replied:
Also not knowing the inner-workings of Flow: does it change anything if the games and runs are resources owned by the admin wallet (instead of a struct)? Then when a tx needs to fetch details/scores about the game we would do a borrow operation on a specific game.

joshuahannan replied:
yeah, probably does make sense to just store them centrally.

HouseOfHufflepuff replied:
Resolving this convo as it appears a central ledger is operative here for the time being.

turbolent commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+        ///
+        access(contract) fun incrementLeaderboardWins(wallet: Address) {
+            let leaderboard = self.leaderboard
+            var wins = leaderboard[wallet] ?? 0
+            wins = wins + 1
+            leaderboard[wallet] = wins
```
```suggestion
            leaderboard[wallet] = (leaderboard[wallet] ?? 0) + 1
```

turbolent commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+
+    /// Named Paths
+    ///
+    pub let CollectionStoragePath:  StoragePath
```
Here and below: whitespace is off

turbolent reviewed (commented):
Great work!

Reviewed the code purely from a Cadence-code style perspective, I have no idea about the game logic, so cannot say if this works as expected

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+                FastBreak.fastBreakGameByID[fastBreakGameID]!.isPublic: "fast break game is private"
+            }
+
+            let acct = getAccount(self.owner?.address!)
```
Not sure if this can really be a constant because the owner can change if the resource is moved to a different account and there wouldn't be a way to update it

turbolent replied:
I see 👍 So depending on which variant is correct, either the initial owner keeps on being used, or always the current owner at each step of the function, maybe make it explicit that the code has that intent.

HouseOfHufflepuff replied:
Added context and intention into the comments of this function. Thanks.

tpetrychyn commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+                valueNeeded: valueNeeded
+            )
+
+            fastBreakGame.addStat(stat: fastBreakStat)
```
Do we need to leave open the possibility of deleting or modifying a stat in the event of an incorrect stat getting added?

HouseOfHufflepuff replied:
Currently this is a no in the off-chain game because it would strand wallets who already submitted to the fast break.

One change I want to make in the short term is to allow edits until the Fast Break is in an open state. But this may not ever get prioritized, so for the moment, you cannot edit stats once a fast break has been created. Must do so carefully like when we make sets & series.

tpetrychyn commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+
+        /// Updates the submission scores of a Fast Break
+        ///
+        pub fun updateFastBreakScore(fastBreakGameID: String, wallet: Address, points: UInt64, win: Bool) {
```
Any considering needed for the new tie-break rule of lowest serial numbers? Or is that something that will be reconciled off-chain?

HouseOfHufflepuff replied:
Good call. I think that's a good future function because right now, the off-chain version does not allow this because if anyone submitted to the former stat, their record would get stranded.

To unblock that, I need to add a new state to the game like "DRAFT" and allow changes during this period while also restricting user submission when the Fast Break is in this state.

At this point, we don't support this off-chain so need to have the on-chain stats be immutable as well.

Perhaps a function to remove a Fast Break entirely on-mistake would be nice to have.

Deewai commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+        /// Get a wallet's active Fast Break Submission
+        ///
+        pub fun getFastBreakSubmissionByWallet(wallet: Address): FastBreak.FastBreakSubmission? {
+            let fastBreakSubmissions = self.submissions!
```
can you explain why force-unwrap is bad?

joshuahannan replied:
If the value that you're trying to unwrap is `nil`, then the force-unwrap will fail and the entire transaction execution will also fail with an unhelpful error message about the force unwrap failing, but no context. Force unwrap should really only be used when you are 100% sure that the optional will never be `nil`. 

Alternatively, you should use if-let to gracefully unwrap the optional and if it is `nil`, you can handle that without having to fail the entire transaction. This example, a getter function, should just return `nil` if `self.submissions` is `nil` to indicate that the submissions was `nil`. As a last resort, if you have to revert the transaction if the value is `nil`, you should just `panic()` with a descriptive error message, because that is much more helpful than just failing with none

joshuahannan replied:
@HouseOfHufflepuff reminder about this

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+    pub event ContractInitialized()
+    pub event Withdraw(id: UInt64, from: Address?)
+    pub event Deposit(id: UInt64, to: Address?)
+    pub event FastBreakRunCreated(
```
I would recommend emitting the UUIDs for all the NFT in these events as well

HouseOfHufflepuff replied:
Am emitting this event when the game token is minted. Currently line 349. LMK if this is not what you meant.

emit FastBreakGameTokenMinted(
    id: self.id,
    fastBreakGameID: self.fastBreakGameID,
    serialNumber: self.serialNumber,
    mintingDate: self.mintingDate,
    topShots: self.topShots,
    mintedTo: self.mintedTo
)

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+import NonFungibleToken from 0xNFTADDRESS
+import TopShot from 0xTOPSHOTADDRESS
+
+/// Game & Oracle Contract for Fast Break
```
Valiant effort on the comments, but what I meant when I asked for comments was to have comments that actually give some context on what things are for. like for each event have a description of the specific context when the event is emitted, or for a type definition, what the type is for and how it is used in practice, and comments for the unclear fields

Additionally, some comments at the top of the file defining some common terminology, like "run", "game", "fatigue", etc

HouseOfHufflepuff replied:
Absolutely, added depth to comments, especially field and game definitions.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+    access(self) let fastBreakRunByID:        {String: FastBreakRun}
+    access(self) let fastBreakGameByID:           {String: FastBreakGame}
+
+    /// A top-level Fast Break Run, the container for Fast Break Games
```
What do you mean by "container for fast break games"

HouseOfHufflepuff replied:
Will make the comment better, but a Fast Break occurs on any day there are NBA games. A Fast Break Run is a collection of Fast Breaks spanning about 2 weeks. We have rewards setup for hitting thresholds. 

<img width="458" alt="Screen Shot 2023-12-21 at 12 18 59 PM" src="https://github.com/dapperlabs/nba-smart-contracts/assets/896651/b0c11549-83da-40e4-8b82-7c779b711b5b">

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+
+        /// Write a new win to the Fast Break Run leaderboard
+        ///
+        access(contract) fun incrementLeaderboardWins(wallet: Address) {
```
probably should use a different name than `wallet`, like `account`, or something. It might also be better to tie stats to resource IDs instead of accounts too because that is more accurate to the cadence programming model

joshuahannan replied:
It also would allow people to trade and sell their fastbreak player records since the records are tied to NFTs instead of accounts, so it is another feature that takes advantage of web3 unique features

HouseOfHufflepuff replied:
I have renamed wallet to account. Good call.

I gave the concept of using a resourceId a lot of thought and it has advantages like you describe. I never considered that originally, so it was fun to think through. The reason I chose a flow account address here is two-fold. First, I made the off-chain version first and I need to tie non-custodial submissions to something I know about off-chain which is flowAddress. Next, this game token is "soul-bound" lite in that it records the address of who won it, but does not restrict ownership like a true soul-bound token would. There are a lot of opinions on this and have run this concept up to Dete for a related project and this is his preferred "trophy" implementation. Record the address of the winner and keep the token an NFT. People building on this can choose to use or discard this additional data of the token.

I do dig the concept of resourceIds here but our systems are built around flow address and am going to go this route to make integration easier.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+
+    /// Metadata Dictionaries
+    ///
+    access(self) let fastBreakRunByID:        {String: FastBreakRun}
+    access(self) let fastBreakGameByID:           {String: FastBreakGame}
```
Any reason the IDs are strings and not integers?

HouseOfHufflepuff replied:
Artifact of doing the off-chain version first. I used strings there.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+            self.points = 0
+            self.win = false
+
+            emit FastBreakSubmissionSent(
```
agreed, we shouldn't do any state changing operations in struct `init` blocks

joshuahannan reviewed (commented):
Please correct me if I am wrong, but it looks like the actual NFTs that you are defining here are just meant to only be used in a single game and are not used any time afterwards? If that is the case, it doesn't really make sense to make them the NFTs and it would make a lot more sense to make the NFT represent the player instead of the game, so it can be used indefinitely across multiple different games, can be moved to a new address if the player wants to for whatever reason, can be traded or sold on marketplaces, and can contain meaningful metadata about the player' history. 

In addition, you'll also need to have this contract implement `ViewResolver` and use metadata views for the NFTs and the contract

HouseOfHufflepuff commented on transactions/fastbreak/oracle/add_stat_to_game.cdc:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    post {
+        FastBreak.getFastBreakGameStats(id: fastBreakGameID).length != 0: "could not find fast break stat"
```
Good call. Removed this from tx in favor of a script & test.

HouseOfHufflepuff commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+                self.isPublic = isPublic
+                self.submissionDeadline = submissionDeadline
+                self.numPlayers = numPlayers
+                self.status = "FAST_BREAK_OPEN"
```
Good call. Replaced with an enum.

HouseOfHufflepuff commented on transactions/fastbreak/user/play.cdc:
```diff
@@ -0,0 +1,30 @@
+            fastBreakGameID: fastBreakGameID,
+            topShots: topShots
+        )
+        self.recipient.deposit(token: <- (nft as @NonFungibleToken.NFT))
```
I gave it a try and it looks like it is needed.

HouseOfHufflepuff commented on lib/go/contracts/contracts.go:
```diff
@@ -27,6 +27,7 @@ const (
 	defaultTopShotLockingAddress   = "TOPSHOTLOCKINGADDRESS"
 	defaultTopShotRoyaltyAddress   = "TOPSHOTROYALTYADDRESS"
 	defaultNetwork                 = "${NETWORK}"
+	fastBreakFile                  = "FastBreak.cdc"
```
Created an issue for this to be prioritized.

https://github.com/dapperlabs/nba-smart-contracts/issues/236

HouseOfHufflepuff commented:
> Please correct me if I am wrong, but it looks like the actual NFTs that you are defining here are just meant to only be used in a single game and are not used any time afterwards? If that is the case, it doesn't really make sense to make them the NFTs and it would make a lot more sense to make the NFT represent the player instead of the game, so it can be used indefinitely across multiple different games, can be moved to a new address if the player wants to for whatever reason, can be traded or sold on marketplaces, and can contain meaningful metadata about the player' history.
> 
> In addition, you'll also need to have this contract implement `ViewResolver` and use metadata views for the NFTs and the contract

Yes, the intention is to create a game token that represents a single play of Fast Break. I chose this design because it maximizes the flexibility other contracts can have building on-top of Fast Break.

Having said that, I really like the idea of having a FastBreakPlayer resource or NFT. I think it adds a ton to the game. I started off creating it as a resource over here, but I am thinking it might be better as an NFT in its own contract. Would appreciate advice or collaboration on this PR on the FastBreakPlayer concept.

https://github.com/dapperlabs/nba-smart-contracts/pull/237

tpetrychyn reviewed (approved):
Reviewed via a slack huddle, LGTM as a v1

tpetrychyn commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,718 @@
+
+            /// Must own Top Shots to play Fast Break
+            for flowId in topShots {
+                if !collectionRef.getIDs().contains(flowId) {
```
`getIDs()` gets all of the nft keys as an array from the collection, I wonder if there could be issue iterating over an array the length of the users collection `O(topShots*collectionSize)` times.

One suggestion is attempt to borrow each nft `collectionRef.borrowNFT(id: flowId)` which is O(1) map lookup

HouseOfHufflepuff replied:
Good call out. I thought it was a map for some reason. Updated to your suggestion for the reasons you mentioned. Thank you.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,708 @@
+        access(contract) fun updateScore(accountAddress: Address, points: UInt64, win: Bool): Bool {
+            var isNewWin = false
+            let submissions = self.submissions
+            let submission: FastBreak.FastBreakSubmission = submissions[accountAddress]!
```
Remember to use `if...let` instead of force unwrap

HouseOfHufflepuff replied:
I think I got all of these, thanks Josh.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+    /// Validate Fast Break Submission
+    ///
+    pub fun isValidSubmission(submissionDeadline: UInt64): Bool {
+        if submissionDeadline > UInt64(getCurrentBlock().timestamp) {
+            return true
+        }
+
+        return false
```
@HouseOfHufflepuff reminder on this

joshuahannan replied:
Still need to address this @HouseOfHufflepuff

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,682 @@
+    /// Get the game stats of a Fast Break
+    ///
+    pub fun getFastBreakGameStats(id: String): [FastBreak.FastBreakStat] {
+        let fastBreak = FastBreak.getFastBreakGame(id: id)!
```
@HouseOfHufflepuff reminder on this. If any getter function can not access the information that is being queried, it shouldn't panic. It should either return `nil` or an empty version of the thing being queried. I would also take a look at all the other getter functions in this contract to make sure they all follow that pattern

joshuahannan reviewed (commented):
Just left a few reminders on some of bastians original review comments. Lets get those fixed now, and I'll be able to get to a full review soon

joshuahannan reviewed (commented):
So I take it that you don't want to made the player a resource instead of tracking things by account address? It doesn't seem like you made any significant changes since the last time I looked at this

HouseOfHufflepuff commented:
@joshuahannan, I want to make the Player an NFT in a new contract and started working on it in the PR mentioned in my comment. Having said that, I do think a game token adds more to the goals of the game, is more experimental, and more composable, and I am planning to stick with this concept. However, I do like the idea of a player NFT as well, but correct me if I am wrong, that would need to be in a different contract. 

Either way, the core functionality I want in this contract is present in this PR. Have made all the suggestions Bastian made afaik. Thanks again for your time. 

https://github.com/dapperlabs/nba-smart-contracts/pull/229#issuecomment-1871637570

joshuahannan commented:
oh sorry, I think I missed that last part of your comment last week. I disagree that you are able to implement the player as a resource in a different contract. You're already tracking players by address in this contract, so if you were to try to graft a resource from a separate contract onto this as a player, it wouldn't work IMO because it can never be reconciled with the address tracking that you're doing here. It is a fundamental enough part of the game that you would need to track players by only their resource IDs and remove address tracking completely. That is also a more Cadence-best practices way to do it.

HouseOfHufflepuff commented:
Hey @joshuahannan, sorry for the delay as I am working on this over weekends. I implemented the Player resource and removed wallet data in favor of a resourceId. I like the concept a lot and feel like this really improved the contract. Thanks for pushing me to do this. Would appreciate your re-review.

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,802 @@
+    /// Resource for playing Fast Break
+    /// The Fast Break Player plays the game & mints game tokens
+    ///
+    pub resource Player: FastBreakPlayer {
```
```suggestion
    pub resource Player: FastBreakPlayer, NonFungibleToken.INFT {
```
I think it makes sense to make this an NFT also. No need to have a collection defined for it too, though, as long as it implements the interface, that should be fine, especially once Cadence 1.0 comes around

joshuahannan commented on contracts/FastBreak.cdc:
```diff
@@ -0,0 +1,802 @@
+            pre {
+                FastBreak.fastBreakGameByID.containsKey(fastBreakGameID): "no such fast break game"
+            }
+
```
Here is where you could put the update to the mapping:
```suggestion
// updating player address mapping
if let ownerAddress = self.owner?.address {
    FastBreak.playerAccountMapping[self.id] = ownerAddress
}
```

joshuahannan reviewed (approved):
Looks good. I like the player resource model better. One thing you do need to be aware of with this model is that because most web3 people still care so much about addresses, people are probably still going to want to be able to see stats based on addresses instead of player IDs because they're stuck in the old ways and don't understand that tying things to resources is much better in terms of composability and extensibility. unfortunately, the'll want it either way.

There are a few ways you could handle this. You could just say no and tell people to adjust to this new model. That might work, but you would need to push this model pretty heavily in the design of the user facing stats and game. You could also emit addresses in events and such like we do with `Withdraw` and `Deposit` to at least have the events include an address. You just can get that from a player or collection resource by doing `self.owner?.address`.

Or, you could make all this information more accessible on-chain by having a mapping of playerIDs to addresses to show which address owns which player at any given time. You could keep this updated by just putting a line in one of the player methods that just keeps it updated automatically.
